gossip/src/push_active_set.rs

```diff
@@ -0,0 +1,316 @@
+
+pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+
+// Keys are gossip nodes to push messages to.
+// Values are which origins the node has pruned.
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get_nodes<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
+        self.get_entry(stake).get_nodes(origin, should_force_push)
+    }
+
+    pub(crate) fn prune(
+        &self,
+        pubkey: &Pubkey,    // This node.
+        node: &Pubkey,      // Gossip node.
+        origins: &[Pubkey], // CRDS value owners.
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let stake = stakes.get(pubkey);
+        for origin in origins {
+            if origin == pubkey {
+                continue;
+            }
+            let stake = stake.min(stakes.get(origin));
+            self.get_entry(stake).prune(node, origin)
+        }
+    }
+
+    pub(crate) fn rotate<R: Rng>(
+        &mut self,
+        rng: &mut R,
+        size: usize, // Number of nodes to retain in each active-set entry.
+        cluster_size: usize,
+        nodes: &[Pubkey],
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let num_bloom_filter_items = cluster_size.max(Self::MIN_NUM_BLOOM_ITEMS);
+        let buckets: Vec<_> = nodes
+            .iter()
+            .map(|node| get_stake_bucket(stakes.get(node)))
+            .collect();
+        for (k, entry) in self.0.iter_mut().enumerate() {
+            let weights: Vec<u64> = buckets
+                .iter()
+                .map(|&bucket| {
+                    // bucket <- get_stake_bucket(min stake of {
+                    //  this node, crds value owner and gossip peer
+                    // })
+                    // weight <- (bucket + 1)^2
+                    // min stake of {...} is a proxy for how much we care about
+                    // the link, and tries to mirror similar logic on the
+                    // receiving end when pruning incoming links:
+                    // https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105
+                    let bucket = bucket.min(k) as u64;
+                    bucket.saturating_add(1).saturating_pow(2)
+                })
+                .collect();
+            entry.rotate(rng, size, num_bloom_filter_items, nodes, &weights);
```

Comment:
Question since the old rotation code is quite hard to read. How does this new rotation logic compare/is it significantly different?

Reply:
Instead of these weights:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L465-L471
I am using weights as calculated above.

When discarding entries:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L418-L420
I am preserving the order of nodes in `IndexMap`.

Otherwise should be functionally the same.

Reply:
got it, yeah that old code based on when the node was last pushed to seems kind of hacky 😝